ios/engine/KMEI/KeymanEngine/Classes/KeymanInputViewController.swift

```diff
@@ -0,0 +1,308 @@
+    }
+    super.updateViewConstraints()
+  }
+
+  open override func viewDidLoad() {
+    super.viewDidLoad()
+
+    let bgColor = UIColor(red: 210.0 / 255.0, green: 214.0 / 255.0, blue: 220.0 / 255.0, alpha: 1.0)
+    view.backgroundColor = bgColor
+
+    Manager.shared.inputViewDidLoad()
+    Manager.shared.inputDelegate = self
+
+    topBarImageView?.removeFromSuperview()
+    topBarImageView = UIImageView()
+    topBarImageView!.translatesAutoresizingMaskIntoConstraints = false
+    topBarImageView!.backgroundColor = UIColor.gray
+    view.addSubview(topBarImageView!)
+
+    containerView?.subviews.forEach { $0.removeFromSuperview() }
+    containerView?.removeFromSuperview()
+    containerView = UIView()
+    containerView!.translatesAutoresizingMaskIntoConstraints = false
+    containerView!.backgroundColor = bgColor
+    containerView!.addSubview(kmInputView)
+    view.addSubview(containerView!)
+  }
+
+  open override func viewDidAppear(_ animated: Bool) {
+    super.viewDidAppear(animated)
+    setConstraints()
+    inputView?.setNeedsUpdateConstraints()
+  }
+
+  open override func willRotate(to toInterfaceOrientation: UIInterfaceOrientation, duration: TimeInterval) {
+    Manager.shared.inputViewWillRotate(to: toInterfaceOrientation, duration: duration)
+  }
+
+  open override func textDidChange(_ textInput: UITextInput?) {
+    let contextBeforeInput = textDocumentProxy.documentContextBeforeInput ?? ""
+    let contextAfterInput = textDocumentProxy.documentContextAfterInput ?? ""
+    let context = "\(contextBeforeInput)\(contextAfterInput)"
+
+    let newRange: Range<String.Index>
+    if let range = context.range(of: contextBeforeInput) {
+      newRange = range.upperBound..<range.upperBound
+    } else {
+      newRange = context.startIndex..<context.startIndex
+    }
+
+    Manager.shared.setText(context)
+    Manager.shared.setSelectionRange(NSRange(newRange, in: context), manually: false)
+  }
+
+  private func processInsertText(_ fragment: String) {
+    if Manager.shared.isSubKeysMenuVisible {
+      return
+    }
+
+    if isInputClickSoundEnabled {
+      DispatchQueue.global(qos: .default).async { AudioServicesPlaySystemSound(0x450) }
+
+      // Disable input click sound for 0.1 second to ensure it plays for single key stroke.
+      isInputClickSoundEnabled = false
+      perform(#selector(self.enableInputClickSound), with: nil, afterDelay: 0.1)
+    }
+
+    let dnRange = fragment.range(of: "+dn=")!
+    let sRange = fragment.range(of: "+s=")!
+
+    let dn = Int(fragment[dnRange.upperBound..<sRange.lowerBound])!
+    let s = fragment[sRange.upperBound...]
+
+    let hexStrings: [String]
+    if !s.isEmpty {
+      hexStrings = s.components(separatedBy: ",")
+    } else {
+      hexStrings = []
+    }
+
+    let codeUnits = hexStrings.map { hexString -> UInt16 in
+      let scanner = Scanner(string: hexString)
+      var codeUnit: UInt32 = 0
+      scanner.scanHexInt32(&codeUnit)
+      return UInt16(codeUnit)
```

Comment:
`scanHexInt32` is native Swift. Before it was `scanHexInt`. Don't think there was an ObjC extension before, but I did add a new extension to `String` that does this conversion in `String+Helpers.swift` although I think this portion was converted before I added in that extension.

I'm also working on abstracting out KMW so that all this conversion can happen in one place. Also since we only use `WKWebKit` now, we can actually send a JavaScript object instead of constructing a string, so that could help with the parsing.